pkg/hub/submarinerconfig/controller.go

```diff
@@ -0,0 +1,269 @@
+				break
+			}
+		}
+		if !hasFinalizer {
+			config.Finalizers = append(config.Finalizers, submarinerConfigFinalizer)
+			_, err := c.configClient.SubmarineraddonV1alpha1().SubmarinerConfigs(config.Namespace).Update(ctx, config, metav1.UpdateOptions{})
+			return err
+		}
+	}
+
+	// config is deleting, we remove its related resources
+	if !config.DeletionTimestamp.IsZero() {
+		if err := c.cleanUpSubmarinerClusterEnv(ctx, config); err != nil {
+			return err
+		}
+		return c.removeConfigFinalizer(ctx, config)
+	}
+
+	if config.Spec.CredentialsSecret == nil {
+		// no platform credentials, the submariner cluster environment neet not to be prepared
+		return nil
+	}
+
+	managedCluster, err := c.clusterClient.ClusterV1().ManagedClusters().Get(context.TODO(), config.Namespace, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		// related managed cluster is not found, we cannot prepare the submariner cluster environment for it
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// prepare submariner cluster environment
+	errs := []error{}
+	cloudProvider, preparedErr := cloud.GetCloudProvider(c.kubeClient, c.workClient, c.eventRecorder, managedCluster, config)
+	if preparedErr == nil {
+		preparedErr = cloudProvider.PrepareSubmarinerClusterEnv()
+	}
+
+	condition := metav1.Condition{
+		Type:    configv1alpha1.SubmarinerConfigConditionEnvPrepared,
+		Status:  metav1.ConditionTrue,
+		Reason:  "SubmarinerClusterEnvPrepared",
+		Message: "Submariner cluster environment was prepared",
+	}
+	switch {
+	case preparedErr == nil:
+		// add a finalizer to managed cluster to avoid the managed cluster is deleted in advance,
+		// because when we clean up the submariner cluster environment, we need the cluster informations
+		if err := c.addManagedClusterFinalizer(ctx, managedCluster); err != nil {
+			return err
+		}
+	case preparedErr != nil:
+		condition.Status = metav1.ConditionFalse
+		condition.Reason = "SubmarinerClusterEnvPreparationFailed"
+		condition.Message = fmt.Sprintf("Failed to prepare submariner cluster environment: %v", preparedErr)
+		errs = append(errs, preparedErr)
+	}
+
+	_, updated, updatedErr := helpers.UpdateSubmarinerConfigStatus(
+		c.configClient,
+		config.Namespace, config.Name,
+		helpers.UpdateSubmarinerConfigConditionFn(condition),
+	)
+	if updatedErr != nil {
+		errs = append(errs, updatedErr)
+	}
+	if updated {
+		c.eventRecorder.Eventf("SubmarinerClusterEnvPrepared", "submariner cluster environment was prepared for manged cluster %s", config.Namespace)
+	}
+
+	return operatorhelpers.NewMultiLineAggregate(errs)
+}
+
+func (c *submarinerConfigController) cleanUpSubmarinerClusterEnv(ctx context.Context, config *configv1alpha1.SubmarinerConfig) error {
+	// no platform credentials, the submariner cluster environment is not prepared
+	if config.Spec.CredentialsSecret == nil {
+		return nil
+	}
+	managedCluster, err := c.clusterClient.ClusterV1().ManagedClusters().Get(ctx, config.Namespace, metav1.GetOptions{})
```

Comment:
what  if mannagedCluster does not exist?

Reply:
after using status instead of finalizer, this will not need